File: Problems/704_binarySearch.ts
function search(nums: number[], target: number): number {
    //use binary search to find target in nums
    //two pointers: left and right

    let [left, right] = [0, nums.length - 1];
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const midNumber = nums[mid];
        
        if (target == midNumber) return mid;
        else if (target > midNumber) left = mid + 1;
        else right = mid - 1;
    }

    return -1;
};

console.log(search([-1,0,3,5,9,12], 2))

File: Problems/704_binarySearch.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        
        #use binary search to find target
        left = 0
        right = len(nums) - 1 #[left, right] represents the current interval being searched

        while left < right:
            midpoint = (left + right) // 2
            
            if target > nums[midpoint]: #discard left side of list
                left = midpoint + 1
            else: #target <= nums, so we discard right side but keep midpoint
                right = midpoint
            
        #left == right, so check if index is correct
        return left if nums[left] == target else -1
